src/dev.rs

```diff
@@ -0,0 +1,199 @@
+    /// A gate was not satisfied for a particular row.
+    Gate { gate_index: usize, row: usize },
+    /// A lookup input did not exist in its corresponding table.
+    Lookup { lookup_index: usize, row: usize },
+}
+
+/// A test
+pub struct MockProver<F: Group> {
+    n: u32,
+    domain: EvaluationDomain<F>,
+    cs: ConstraintSystem<F>,
+
+    // The fixed cells in the circuit, arranged as [column][row].
+    fixed: Vec<Polynomial<F, LagrangeCoeff>>,
+    // The advice cells in the circuit, arranged as [column][row].
+    advice: Vec<Polynomial<F, LagrangeCoeff>>,
+    // The aux cells in the circuit, arranged as [column][row].
+    aux: Vec<Polynomial<F, LagrangeCoeff>>,
+
+    permutations: HashMap<usize, Vec<(Cell, Cell)>>,
+}
+
+impl<F: Field + Group> Assignment<F> for MockProver<F> {
+    fn assign_advice(
+        &mut self,
+        column: crate::plonk::Column<crate::plonk::Advice>,
+        row: usize,
+        to: impl FnOnce() -> Result<F, crate::plonk::Error>,
+    ) -> Result<(), crate::plonk::Error> {
+        *self
+            .advice
+            .get_mut(column.index())
+            .and_then(|v| v.get_mut(row))
+            .ok_or(Error::BoundsFailure)? = to()?;
+
+        Ok(())
+    }
+
+    fn assign_fixed(
+        &mut self,
+        column: crate::plonk::Column<crate::plonk::Fixed>,
+        row: usize,
+        to: impl FnOnce() -> Result<F, crate::plonk::Error>,
+    ) -> Result<(), crate::plonk::Error> {
+        *self
+            .fixed
+            .get_mut(column.index())
+            .and_then(|v| v.get_mut(row))
+            .ok_or(Error::BoundsFailure)? = to()?;
+
+        Ok(())
+    }
+
+    fn copy(
+        &mut self,
+        permutation: usize,
+        left_column: usize,
+        left_row: usize,
+        right_column: usize,
+        right_row: usize,
+    ) -> Result<(), crate::plonk::Error> {
+        self.permutations
+            .entry(permutation)
+            .or_default()
+            .push((Cell(left_column, left_row), Cell(right_column, right_row)));
+        Ok(())
+    }
+}
+
+impl<F: FieldExt> MockProver<F> {
+    pub fn run<ConcreteCircuit: Circuit<F>>(
+        k: u32,
+        circuit: &ConcreteCircuit,
+        aux: Vec<Polynomial<F, LagrangeCoeff>>,
+    ) -> Self {
+        let mut cs = ConstraintSystem::default();
+        let config = ConcreteCircuit::configure(&mut cs);
+
+        // The permutation argument will serve alongside the gates, so must be
+        // accounted for.
+        let mut degree = cs
+            .permutations
+            .iter()
+            .map(|p| p.required_degree())
+            .max()
+            .unwrap_or(1);
+
+        // Account for each gate to ensure our quotient polynomial is the
+        // correct degree and that our extended domain is the right size.
+        for poly in cs.gates.iter() {
+            degree = std::cmp::max(degree, poly.degree());
+        }
+
+        let domain = EvaluationDomain::new(degree as u32, k);
```

Comment:
Currently nothing in `MockProver` actually relies on this, so if neither the permutation check nor anything else needs it, I can remove it to speed up tests.

Reply:
Yep, from the circuit's perspective the domain should be irrelevant due to the `Rotation` API keeping the actual points in the domain opaque. You can remove this.